Goldtree/Goldtree/Program.cs

```diff
@@ -86,204 +83,228 @@ public static void Initialize()
-                                                {
-                                                    Log.Log("Sending certificate file...");
-                                                    PfsFileEntry cert = pnsp.Files[certidx];
-                                                    br.BaseStream.Seek(pnsp.HeaderSize + cert.Offset, SeekOrigin.Begin);
-                                                    byte[] file = br.ReadBytes((int)cert.Size);
-                                                    USB.Write(file);
-                                                    Log.Log("Certificate was sent to Goldleaf.");
-                                                }
-                                                else if(ccmd.IsCommandId(CommandId.Finish)) break;
-                                                else
-                                                {
-                                                    USB = null;
-                                                    Error.Log("An invalid command was received. Are you sure Goldleaf is active?");
-                                                }
-                                            }
-                                            else
-                                            {
-                                                USB = null;
-                                                Error.Log("An invalid command was received. Are you sure Goldleaf is active?");
-                                            }
-                                        }
-                                    }
-                                    catch
-                                    {
-                                        Error.Log("An error ocurred opening the selected NSP. Are you sure it's a valid NSP?");
-                                    }
-                                }
-                                else if(rc2.IsCommandId(CommandId.Finish))
-                                {
-                                    USB = null;
-                                    Error.Log("Goldleaf has canceled the installation.");
-                                }
-                                else
-                                {
-                                    USB = null;
-                                    Error.Log("An invalid command was received. Are you sure Goldleaf is active?");
-                                }
-                            }
-                            else
-                            {
-                                USB = null;
-                                Error.Log("An invalid command was received. Are you sure Goldleaf is active?");
-                            }
+                            if (rsize >= ent.Size) rsize = ent.Size;
+                            int tor = (int)Math.Min(rsize, toread);
+                            bufb = new byte[tor];
```

Comment:
This method is rough, could use some refactor. Variable names could be more helpful too.